### ISSUE TITLE
stats: contention on stats table causes slow SHOW TABLES command

### ISSUE THREAD

**Issue by rytaft** [cockroachdb/cockroach#58189]
We have heard reports that contention on the stats table is causing the `SHOW TABLES` command to execute very slowly. In 20.2 we augmented the `SHOW TABLES` command to include information collected from stats collection. This turns out to be a performance regression, especially in multi-region clusters as fetching the stats requires a lookup. This alone isn't always so bad as each lookup should be quick, however, sometimes the lookup encounters contention with a job writing new stats. The process of writing new stats happens incrementally and in a long-running cluster can take seconds. 

Users of CRDB who are running into this issue today on 20.2.x can use the following workarounds:
1. Run your `SHOW TABLES` at `PRIORITY HIGH`. This will require using an explicit transaction. For example: `BEGIN PRIORITY HIGH; SHOW TABLES; COMMIT;`.
2. Reduce the frequency of stats refreshes to make contention less common. As described [here](https://www.cockroachlabs.com/docs/stable/cost-based-optimizer.html#controlling-statistics-refresh-rate), you can do this by tuning the cluster settings `sql.stats.automatic_collection.fraction_stale_rows` and `sql.stats.automatic_collection.min_stale_rows`. There is a potential downside with this approach, though, since it could mean that the stats are more stale and therefore the optimizer might choose a worse plan.

---

For future releases and/or backports, we should try to reduce contention on the stats table so the above workarounds aren't necessary.
As suggested by @ajwerner, we can reduce contention when stats are written as follows:
1. Run the txn at low priority, it's bad for creating stats to block running transactions or show tables
2. Generate the encoded data above the txn
3. Batch the data insertion as it's probably all going to the same range anyway

Another suggestion is to support `WITHOUT STATISTICS` as a modifier to `SHOW TABLES`, which will make the query much faster regardless of contention.

Finally, we might consider using the stats cache for `SHOW TABLES`. We're currently querying the system table directly, but we could probably avoid some lookups if we use the cache. One possible issue with this approach is that if the cache is full, it will cause entries to be evicted by table stats that may not be useful for other queries. We could write a new function `StealthyGetTableStats`, which would in turn use the function `StealthyGet` that @RaduBerinde recently added to avoid counting a `Get` towards cache eviction.

**Comment by RaduBerinde** [comment]
Could SHOW TABLES query that table using `AS OF SYSTEM TIME -10s`?

**Comment by rytaft** [comment]
> Could SHOW TABLES query that table using `AS OF SYSTEM TIME -10s`?

That's a good idea -- that would be a simple backport, along with support for `WITHOUT STATISTICS`.

**Comment by ajwerner** [comment]
> > Could SHOW TABLES query that table using `AS OF SYSTEM TIME -10s`?
> 
> That's a good idea -- that would be a simple backport, along with support for `WITHOUT STATISTICS`.

We don't currently support running different portions of queries at different timestamps. `SHOW TABLES` is transactional and I'd be very afraid to break that property.

**Comment by RaduBerinde** [comment]
Ah, it's all a big query..
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/delegate/show_tables.go#L55

**Comment by ajwerner** [comment]
We may want to punt part of this to @cockroachdb/sql-experience to strip the stats off of `SHOW TABLES` optionally and to litigate what the default should be. Putting it on their board too.

**Comment by rafiss** [comment]
i would be in favor of vanilla `SHOW TABLES` being stripped down to only include `schema_name, table_name, type, owner`. (postgres doesn't have a SHOW command, but this matches the output of the `\d` metacommand from psql).

**Comment by ajwerner** [comment]
> i would be in favor of vanilla `SHOW TABLES` being stripped down to only include `schema_name, table_name, type, owner`. (postgres doesn't have a SHOW command, but this matches the output of the `\d` metacommand from psql).

I absolutely support that for the next major version but I don't think we can backport that.

**Comment by rafiss** [comment]
For backport purposes, we'll probably address this by adding a cluster setting that disables showing the row count in SHOW TABLES output.

**Comment by ajwerner** [comment]
[This suggestion](https://github.com/cockroachdb/cockroach/pull/59366#issuecomment-767712735) feels totally backportable and wonderful. It may not solve all use cases where users have code that calls vanilla `SHOW TABLES` and doesn't specify the columns but then again, changing the cluster setting value also may break an application (depending on the driver etc).

**Comment by otan** [comment]
So for 20.2, we want to add a cluster setting that people can set to OFF to avoid displaying this.
What do we want as the behaviour for 21.1?

* Same cluster setting? Default to on or off? If different default to 20.2, people migrating over may feel pain.
* add syntax WITH STATISTICS? There's already WITH COMMENT, what if people want both?

cc @vy-ton

**Comment by vy-ton** [comment]
> i would be in favor of vanilla SHOW TABLES being stripped down to only include schema_name, table_name, type, owner. (postgres doesn't have a SHOW command, but this matches the output of the \d metacommand from psql).

This would be a backwards-incomptabile change that I'm not sure is worth doing. For 20.2 and beyond, a user can set the cluster setting to `OFF` if they want to avoid displaying row count. For future `SHOW` columns, we can consider the `WITH` syntax.

cc @awoods187 for awareness

**Comment by otan** [comment]
What would the WITH syntax look like if you want stats and comments?
```
SHOW TABLES WITH STATISTICS # no comment?
SHOW TABLES WITH COMMENT, STATISTICS
SHOW TABLES WITH COMMENT WITH STATISTICS
```

**Comment by vy-ton** [comment]
One guiding principle could be to only support the `WITH` syntax for columns with a performance cost. So we could add `comment` to the output for `SHOW TABLES`. 

Unless we want to make `SHOW TABLES` declarative and specify every column, I treat `SHOW TABLES = SELECT *`. You can fetch specify columns by using `SELECT _ FROM [SHOW...]`

There's probably more thinking we can do around `SHOW` commands as part of our public API

**Comment by otan** [comment]
I should be clearer, we already have `SHOW TABLES WITH COMMENT`. The question is how COMMENT and STATISTICS should look if we are using `WITH`

**Comment by otan** [comment]
update from meeting with @vy-ton 

We'll continue display the column 21.1, to maintain backwards compatibility. The cluster setting will still be available to deactivate by default. At the same time, we'll introduce WITH STATISTICS, which always displays estimated_row_count (even if the cluster setting is enabled). In 21.2, we'll remove the column showing by default and the cluster setting.

This way, there is a way forward from v20.2 -> v21.1 without breaking clients if they rely on estimated_row_counts as a column.

**Comment by vy-ton** [comment]
> In 21.2, we'll remove the `estimated_row_count` column showing by default and the cluster setting.

@ericharmeling From a documentation perspective, how should we convey this? In 21.1, `SHOW TABLES WITH STATISTICS` is being added for anyone who relies on `estimated_row_count`, so users will have 21.1 to migrate away. In 21.2, `estimated_row_count` will be disabled by default so it would be a backwards-incompatible change unless they switched to using `WITH STATISTICS`

**Comment by ajwerner** [comment]
I still strongly advocate that we do https://github.com/cockroachdb/cockroach/issues/58189#issuecomment-767790744 also. It means that even if you leave the setting on, if you don't select the column, you won't incur the costs. My hunch is that many users which are pragmatically accessing `SHOW TABLES` are doing projections to columns they care about.

**Comment by otan** [comment]
I feel as if having a user having to explicitly do SELECT estimated_row_count FROM [Show tables] makes the UI even less obvious if I understand the suggestion.

**Comment by ajwerner** [comment]
> I feel as if having a user having to explicitly do SELECT estimated_row_count FROM [Show tables] makes the UI worse if I understand the suggestion. The rows not being correct unless projected seems like a misleading thing.

This isn't the suggestion. See the PR. It's just that if you don't project it then the join isn't performed. Check out #59838.

**Comment by rafiss** [comment]
just re-read the initial issue description

> For future releases and/or backports, we should try to reduce contention on the stats table so the above workarounds aren't necessary.
> we can reduce contention when stats are written as follows:
>
>    Run the txn at low priority, it's bad for creating stats to block running transactions or show tables
>    Generate the encoded data above the txn
>    Batch the data insertion as it's probably all going to the same range anyway

if some or all of these improvements to stats writing happen, then do we still need the cluster setting or `WITH STATISTICS`? (maybe they'd be useful as visual UX improvements, but would they be needed for the performance improvement?)

@rytaft were you planning to make those changes to stats writing?

**Comment by rytaft** [comment]
Hi @rafiss, thanks for the reminder. I've had it on my todo list for a while, but things have been a bit crazy...

I agree that these performance improvements would likely help, but I'm not sure they would get rid of the performance problems altogether, since stats updates write a lot of data. But I also don't fully understand how transaction priorities work or how batching changes performance. Maybe @ajwerner can give more insight here. I also don't fully understand the risks of backporting these changes, especially the prioritization change. Could running at low priority cause starvation issues for the process trying to write statistics to the DB?

Either way, I'll try to make these changes to master over the course of the next week.

**Comment by otan** [comment]
Since we have optimisations underway, I've taken sql-experience off this after adding the cluster setting in 21.1: #60282 

I think if I read this correctly there's some optimiser changes that can happen. If we decide we still want to go down the WITH STATISTICS route, feel free to reassign over to us!

**Comment by rafiss** [comment]
@rytaft: I just noticed your comment:
> I also don't fully understand the risks of backporting these changes, especially the prioritization change.

We don't need to backport the changes you make to stats writing -- I was just wondering if the improvements would be in 21.1. The new cluster setting that @otan added is good enough for the backport.

**Comment by awoods187** [comment]
I am still hopeful that we can keep stats in `SHOW TABLES` in 21.1 as I think they are useful

**Comment by rytaft** [comment]
> We don't need to backport the changes you make to stats writing

Great - thanks! I think I should be able to squeeze this stuff in to 21.1

**Comment by RaduBerinde** [comment]
Going back to the idea of AS OF SYSTEM TIME, what would be the problem with issuing the [internal query](https://github.com/cockroachdb/cockroach/blob/74c2a602b5019a8586ae344e155029bbae29bf52/pkg/sql/crdb_internal.go#L431
) for `table_row_statistics` in a separate transaction, with AS OF SYSTEM TIME -10s ?  There is no requirement that the virtual table reflects the statistics as available at the outer txn timestamp.

**Comment by rytaft** [comment]
I just merged #60953 which implements @RaduBerinde's idea from the previous comment. Based on offline discussions, we decided that this idea was preferable to some of the other ideas discussed here such as running the stats writing transaction at low priority (due to the possibility of starvation) and batching of the stats writer SQL statements (it's very difficult and not ergonomic to batch the `DELETE` and `INSERT` SQL statements of the stats writer into a single statement). The stats cache idea was also rejected due to the fact that the cache may not have stats for all tables.
